### ISSUE TITLE
False positive `E0308` in recent versions of rust analyzer

### ISSUE THREAD

**Issue by winderica** [rust-lang/rust-analyzer#21658]
**Context**: We are refactoring a cryptographic library called [`sonobe`](https://github.com/privacy-ethereum/sonobe) in order to make it more generic. This introduces quite complex definitions of traits, and we recently observed `E0308` errors within our codebase after updating rust-analyzer, which were not reported by previous versions of rust-analyzer. These errors turned out to be false positives, as rustc couldn't detect these problems with `cargo check` or `cargo build`.

**rust-analyzer version**: Tested with both `rust-analyzer 0.3.2786-standalone` and `rust-analyzer 0.4.2794-standalone`. I am happy to do a bisection to see this regression is introduced by which version, if necessary.

**rustc version**: Tested with both `rustc 1.93.1 (01f6ddf75 2026-02-11)` and `rustc 1.95.0-nightly (873b4beb0 2026-02-15)`.

**editor or extension**: I am using rust-analyzer as a VSCode extension, but I believe the issue is irrelevant to the editor, because running `rust-analyzer diagnostics` in CLI can produce the same results, and another member of our project is able to reproduce with his setup as well. 

**relevant settings**: I am using default settings and environment variables for rust-analyzer and rustc.

**repository link (if public, optional)**: https://github.com/privacy-ethereum/sonobe/tree/dev

**code snippet to reproduce**: Below is a minimal reproducible example:
```rust
pub trait Field: AdditiveGroup + Default {}

pub trait AdditiveGroup {
    type Scalar: Field;
}

pub trait CommitmentScheme {
    type Commitment;
    type Scalar;
}

pub trait CurveGroup: AdditiveGroup {
    type Base: Field;
}

pub trait GroupBasedCommitmentScheme:
    CommitmentScheme<
        Commitment: CurveGroup<Base: Field>,
        Scalar = <<Self as CommitmentScheme>::Commitment as AdditiveGroup>::Scalar,
    >
{
}

pub fn foo<T>(_x: T, _y: T) {}

pub fn bar<T: GroupBasedCommitmentScheme>(v: T::Scalar) {
    // Error RustcHardError("E0308"): expected <<<T as CommitmentScheme>::Commitment as CurveGroup>::Base as AdditiveGroup>::Scalar, found <<T as CommitmentScheme>::Commitment as AdditiveGroup>::Scalar
    foo(v, <T as CommitmentScheme>::Scalar::default());
}
```

I guess the reason might be due to `AdditiveGroup` and `CommitmentScheme` having associated types of the same name, `Scalar`, which may confuse rust-analyzer, but I am not sure if this is indeed the root cause.

**Comment by ChayimFriedman2** [comment]
Fixed by https://github.com/rust-lang/rust-analyzer/pull/21631 which is already merged!
